jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java

```diff
@@ -350,119 +351,356 @@ public void abort(Throwable failure)
+     * <ul>
+     *   <li>the completion of the send operation, either success or failure</li>
+     *   <li>an asynchronous failure coming from the read side such as a stream
+     *   being reset, or the connection being closed</li>
+     *   <li>an asynchronous idle timeout</li>
+     * </ul>
+     * <p>The last 2 cases may happen <em>during</em> a send, when the frames
+     * are being generated in the flusher.
+     * In such cases, this class must avoid that the nested callback is notified
+     * while the frame generation is in progress, because the nested callback
+     * may modify other states (such as clearing the {@code HttpOutput._buffer})
+     * that are accessed during frame generation.</p>
+     * <p>The solution implemented in this class works by splitting the send
+     * operation in 3 parts: {@code preSend()}, {@code send()} and {@code postSend()}.
+     * Asynchronous state changes happening during {@code send()} are stored
+     * and only executed in {@code postSend()}, therefore never interfering
+     * with frame generation.</p>
+     *
+     * @see State
+     */
     private class TransportCallback implements Callback
     {
-        private State state = State.IDLE;
-        private Callback callback;
-        private Throwable failure;
-        private boolean commit;
+        private State _state = State.IDLE;
+        private Callback _callback;
+        private boolean _commit;
+        private Throwable _failure;
 
-        public boolean start(Callback callback, boolean commit)
+        private void reset()
         {
-            State state;
+            assert Thread.holdsLock(this);
+            _state = State.IDLE;
+            _callback = null;
+            _commit = false;
+            _failure = null;
+        }
+
+        private void send(Callback callback, boolean commit, Consumer<Callback> consumer)
+        {
+            Throwable failure = preSend(callback, commit);
+            if (failure == null)
+            {
+                consumer.accept(this);
+                postSend();
+            }
+            else
+            {
+                callback.failed(failure);
+            }
+        }
+
+        private Throwable preSend(Callback callback, boolean commit)
+        {
+            synchronized (this)
+            {
+                switch (_state)
+                {
+                    case IDLE:
+                    {
+                        _state = State.PRE_SEND;
+                        _callback = callback;
+                        _commit = commit;
+                        return null;
+                    }
+                    default:
+                    {
+                        Throwable failure = _failure;
+                        if (failure == null)
+                            failure = new IllegalStateException("Invalid transport state: " + _state);
+                        return failure;
+                    }
```

Comment:
This is not one case but two. Also needs comments:

```suggestion
                    case FAIL:
                        // better comment that an async failure happened while we were just about to send something
                        return _failure;
                    default:
                        return new IllegalStateException("Invalid transport state: " + _state);
```

Reply:
FAIL here is an illegal state.
FAILED here is legal and I can return `failure` like you propose.
However, is that what you meant?